Update RiverFlood module to GloFAS v4

peanutfun opened the pull request:
Changes proposed in this PR:
- Update to GloFAS v4
- Change handling of flood hazard maps: Download and open individual tiles instead of merging during setup
- Add option to split download requests based on time parameter within the data
- Update docs and tests

This PR fixes #139 #164 #141

### PR Author Checklist

- [ ] Read the [Contribution Guide][contrib]
- [ ] Correct target branch selected (if unsure, select `develop`)
- [ ] Descriptive pull request title added
- [ ] Source branch up-to-date with target branch
- [ ] [Documentation](https://climada-python.readthedocs.io/en/latest/guide/Guide_PythonDos-n-Donts.html#Commenting-&-Documenting) updated
- [ ] [Tests][testing] updated
- [ ] [Tests][testing] passing
- [ ] No new [linter issues][linter]
- [ ] [Changelog](https://github.com/CLIMADA-project/climada_petals/blob/main/CHANGELOG.md) updated

### PR Reviewer Checklist

- [ ] Read the [Contribution Guide][contrib]
- [ ] [CLIMADA Reviewer Checklist](https://climada-python.readthedocs.io/en/latest/guide/Guide_Reviewer_Checklist.html) passed
- [ ] [Tests][testing] passing
- [ ] No new [linter issues][linter]

[contrib]: https://github.com/CLIMADA-project/climada_python/blob/main/CONTRIBUTING.md
[testing]: https://climada-python.readthedocs.io/en/latest/guide/Guide_Continuous_Integration_and_Testing.html
[linter]: https://climada-python.readthedocs.io/en/latest/guide/Guide_Continuous_Integration_and_Testing.html#Static-Code-Analysis

peanutfun committed:
Update EDS requests

peanutfun committed:
Update setup functions to GloFAS v4 data

peanutfun committed:
WIP: Rework setup process

* Download flood maps tiles
* Compute historical discharge and Gumbel fit based on tiles

peanutfun committed:
Add handling of flood maps

Commit:
Add everything used for building Gumbel fit of GloFAS v4

Commit:
Clean up setup.py

peanutfun committed:
Merge branch 'rf-glofas-v4' of https://github.com/CLIMADA-project/climada_petals into rf-glofas-v4

peanutfun committed:
Glofas v4

peanutfun committed:
Merge branch 'develop' into rf-glofas-v4

peanutfun committed:
Replace deprecated assertDictContainsSubset

peanutfun committed:
Merge branch 'rf-glofas-v4' of https://github.com/CLIMADA-project/climada_petals into rf-glofas-v4

luseverin commented:
Thanks a lot @peanutfun for taking the time to update the model! I tried the updated version and it seems to be working quite well so far. I just wanted to report one issue here:
I get a `HTTPError`:
```
HTTPError: 400 Client Error: Bad Request for url: https://ewds.climate.copernicus.eu/api/retrieve/v1/processes/cems-glofas-forecast/execution
invalid request
No valid dates from year=[2024] month=[6] day=[31]
```

Apparently, the issue comes from the way you format the dates in `cds_glofas_downloader.glofas_request`:
```
requests = [
     {**default_request, **dict(zip(split_request_map.keys(), bits))}
     for bits in it.product(*split_request_map.values())
]
```
As it seems to me, doing this you assume that each month has the same number of days, creating the error. A fix could be to catch this error specifically and pass when the requested day does not exist. Otherwise, maybe limit the number of days for unpair months, what do you think?

peanutfun commented:
@luseverin good catch! I think the problem is in `datetime_index_to_request`, where days, months and years are considered separately. Can you tell me exactly which call to `download_glofas_discharge` or `RiverFloodInundation.download_forecast` lead to the error?

luseverin commented:
Thanks for looking into it! I think the following call resulted in the error:
```
start_date = f"{2024}-{3}-{20}"
end_date = f"{2024}-{6}-{10}"

forecast_dates = pd.date_range(start_date, end_date)
lead_time_days = 2
leadtime_hour = list(
    map(str, (np.arange(1, lead_time_days + 1, dtype=np.int_) * 24).flat)
)
#download forecast
forecast_ds = download_glofas_discharge(
    product="forecast",
    dates=forecast_dates,
    countries=["Kenya"],
    preprocess=lambda x:x.max(dim="forecast_period").mean(dim="number", keepdims=True),
    leadtime_hour=leadtime_hour,
    system_version="operational",
    num_proc=8
)
```
Let me know if you manage to replicate this error.

elianekobler commented:
Thanky you @peanutfun for the updated glofas rf model!

After running setup_all() I only got the FLOPROS dataset and the gumbel-fit.nc . However, it is still mentioned in the docstring that setup_all() would download and merge JRC river flood hazard maps. Is the setup_flood_hazard_maps() function excluded from the setup_all() function on purpose?

peanutfun commented:
@elianekobler, yes, this is on purpose. The docs are outdated. The merging of the flood maps is not done anymore because the merged dataset is too large. The individual flood map tiles are downloaded by `RiverFloodInundation` when required.

luseverin commented:
@elianekobler and I have been recurently running into connection error when downloading discharge from the CDS api:
```
WARNING Recovering from connection error [('Connection aborted.', RemoteDisconnected('Remote end closed connection without response'))], attemps 0 of 500
2022-03-18 22:05:30,589 WARNING Retrying in 120 seconds
```
In some cases, it sometimes reconnected after some time but in other cases it was seemingly running forever..
For, me reducing the number of processes (`num_proc=1`) seems to have make things a bit more stable, but it still sometimes occurs. Not sure if this related to unstable connection (however, I've never had this issue with glofas3) or to something else e.g. an excessive number of queries to the server. @peanutfun have you encountered similar issues as well?

peanutfun commented:
@luseverin I frequently encountered timeout issues when downloading the historical data on Euler. It seemed to me like a single process would stall the download when running in parallel. However, I could not narrow the problem down. It might be related to the CDSAPI (after the switch to the new Emergency Data Store), or with running the requests in parallel.

luseverin committed:
Check that dates are valid before sending request

luseverin commented:
> Thanks a lot @peanutfun for taking the time to update the model! I tried the updated version and it seems to be working quite well so far. I just wanted to report one issue here: I get a `HTTPError`:
> 
> ```
> HTTPError: 400 Client Error: Bad Request for url: https://ewds.climate.copernicus.eu/api/retrieve/v1/processes/cems-glofas-forecast/execution
> invalid request
> No valid dates from year=[2024] month=[6] day=[31]
> ```

I added a fix to get rid of this error. However, there is still the issue that the data returned to the user does not match the queried dates (i.e. if you ask for flood forecasts for the period 2024-01-31 to 2024-02-01 you will obtain forecasts for the 2024-01-31 , 2024-02-01  **and** 2024-01-01). This might be problematic especially if one does not check what exactly comes out of the request and directly compute statistic over the entire dataset.

peanutfun commented:
@luseverin regarding the date issue, I did some tests and found out the following:
1. It is impossible to download data for particular days with a months switch in one request: Requesting for 2024-01-31 and 2024-02-01 yields the request 
   ```python
   {"month": ["01", "02"],
    "day": ["31", "01"]}
   ```
   which actually downloads 2024-01-01, 2024-01-31, 2024-02-01, 2024-02-31. Therefore, one always has to split the requests in such a case
2. We do not have to limit the call to `unique` days/months/years because the CDSAPI does so automatically. The following downloads 2024-01-01 amd 2024-01-02:
   ```python
    {"year": ["2024", "2024", "2024"],
     "month": ["01", "01"],
     "day": ["01", "02"]}
   ```

Consequently, we can remove the call to `unique` day/month/year values and make the users aware of the API shortcomings (point 1). With `split_request_keys` it should be possible to find a combination of requests that downloads exactly what you want/need

peanutfun committed:
Split requests based on DatetimeIndex grouping

peanutfun committed:
Remove unused import

peanutfun committed:
Adjust 'split_request' argument in tests

peanutfun committed:
Merge branch 'develop' into rf-glofas-v4

luseverin commented:
Hi @peanutfun, I just had a look and indeed the splitting by day fixes the issue of downloading unwanted dates, well done!

peanutfun committed:
Fix linter issues

* Remove unused imports.
* Add and update docstrings.
* Add module docstrings.
* Rename variables.

peanutfun committed:
Add climada_petals.hazard.rf_glofas.flood_maps to docs

peanutfun committed:
Fix docs and more linter issues

peanutfun committed:
Update tests

peanutfun committed:
Add test_flood_maps

peanutfun committed:
Add option to scale hazard frequency in rf_glofas

Set to True by default.

peanutfun commented:
@luseverin I cannot replicate this error. I just downloaded [the file](https://www.research-collection.ethz.ch/entities/researchdata/be75442a-93d4-4b4e-864c-82bd07c575d7) and I can open it without problems. Can you check if the file is complete? It has a size of 106 MB on my machine. Maybe there is an issue with the download routine for the file.

Please note that this part of the tutorial likely won't work. The forecast from 2021-07-10 was conducted with GloFAS v3, so you will get an error later on when you try to compute the return period. For 2021-07-10, you will have to use reforecast data that was computed with GloFAS v4. I discussed this with @elianekobler yesterday.

peanutfun commented:
The research collection has a new-looking frontend. The download link probably changed. I'll have a look

peanutfun committed:
Update gumbel-fit.nc download link

peanutfun committed:
Update tutorial

peanutfun committed:
Cosmetic updates

peanutfun committed:
Set regridding output chunksizes to reduce memory load

peanutfun commented:
@luseverin The link has changed indeed. I just updated the branch

luseverin commented:
> @luseverin The link has changed indeed. I just updated the branch

Thanks! It works now.

peanutfun committed:
Merge branch 'develop' into rf-glofas-v4

emanuel-schmid committed:
Merge branch 'develop' into rf-glofas-v4

peanutfun commented:
We need to handle the case where the discharge dataset only contains the spatial dimensions. The data pipeline currently expects at least one other dimension.

peanutfun commented:
Todo: Support 2D discharge datasets

peanutfun committed:
Support 2D datasets in `flood_depth`

peanutfun committed:
Merge branch 'develop' into rf-glofas-v4

peanutfun committed:
Fix typing issue in 'flood_depth'

luseverin committed:
Update dates format request to new api specs

luseverin committed:
Fix misaligned tile id for download from JRC

peanutfun commented on climada_petals/hazard/rf_glofas/flood_maps.py:
```diff
@@ -67,7 +67,7 @@ def tile_url(return_period: int, tile_id: int, name: str) -> str:
         The function that loads the tile extents dataset.
     """
-    return JRC_FLOOD_HAZARD_MAP_URL.format(rp=return_period, tile_id=tile_id, name=name)
+    return JRC_FLOOD_HAZARD_MAP_URL.format(rp=return_period, tile_id=tile_id+1, name=name)
```
@luseverin Are you sure about this? The file contents look like this
```
In [2]: gpd.read_file("https://jeodpp.jrc.ec.europa.eu/ftp/jrc-opendata/CEMS-GLOFAS/flood_hazard
      ⋮ /tile_extents.geojson")
Out[2]: 
      id      name                                           geometry
0      1  N70_W180  POLYGON ((-180.00042 70.00207, -170.00125 70.0...
1      2  N80_W170  POLYGON ((-170.00125 80.00123, -160.00208 80.0...
2      3  N70_W170  POLYGON ((-170.00125 70.00207, -160.00208 70.0...
3      4  N60_W170  POLYGON ((-170.00125 60.0029, -160.00208 60.00...
4      5  N80_W160  POLYGON ((-160.00208 80.00123, -150.00292 80.0...
..   ...       ...                                                ...
```
And the file directory looks like this:
<img width="879" height="350" alt="Image" src="https://github.com/user-attachments/assets/4607bf5c-4ad4-4e83-b040-800387c7e4ea" />

Why would you need to increment the tile ID?

peanutfun replied:
I think this was mainly concidence because a new tile was inserted in the beginning. The fix is to update the tests, because the original tiles received new IDs, not to update the link (which will cause the wrong files to be downloaded)

luseverin replied:
Hi @peanutfun thanks for looking into it. Sorry if I applied the wrong fix and thanks for correcting my mistake. Is there anything else which needs to be done? Regarding this specific issue or the update in general.

peanutfun replied:
I think we are good!

peanutfun committed:
Conform GloFAS model to v2.1.2 hazard maps

emanuel-schmid committed:
Merge branch 'develop' into rf-glofas-v4

peanutfun commented:
@emanuel-schmid This seems to be working well for @luseverin and @elianekobler, could you do a review so we can merge this soon and fix the failing compatibility tests from core?

emanuel-schmid commented:
I can already tell so much: the changes in the env_climada.yaml file are not compatible with Windows OS. 😟 
So far the gfortran dependency has been excluded from the environment file for this very reason. Perhaps it is even possible to install gfortran within a conda environment for windows, but certainly not together with the other given dependencies.
Our approach how to deal with was opportunistic. Users who want to use the module need to install gfortran separately, by themselves.
Afaik there is no way to specify optional dependencies in a conda env file, but we could have several os - specific env files.
Until now we didn't want to do that.

Bottom line: we have abandoned fortran for windows and we write the env-file so that it fits all OS.

emanuel-schmid commented on climada_petals/hazard/rf_glofas/setup.py:
```diff
@@ -211,28 +201,32 @@ def download_gumbel_fit(output_dir=DEFAULT_DATA_DIR):
 def setup_all(
     output_dir: Union[str, Path] = DEFAULT_DATA_DIR,
+    compute_gumbel_fit: bool = False,
+    **setup_gumbel_fit_kwargs,
```
name cosmetics: what about `compute_gumbel_fit_kwargs`

peanutfun replied:
Done

emanuel-schmid commented on climada_petals/hazard/rf_glofas/setup.py:
```diff
@@ -211,28 +201,32 @@ def download_gumbel_fit(output_dir=DEFAULT_DATA_DIR):
-    download_gumbel_fit(output_dir)
+
+    if compute_gumbel_fit:
+        setup_gumbel_fit(output_dir, **setup_gumbel_fit_kwargs)
```
name cosmetics: what about `compute_gumbel_fit(...)`  and change the parameter name?

peanutfun replied:
Done

emanuel-schmid commented on climada_petals/hazard/rf_glofas/setup.py:
```diff
@@ -170,37 +97,100 @@ def setup_gumbel_fit(
+    Download dataset of https://doi.org/10.3929/ethz-b-000726304
     """
     LOGGER.debug("Downloading Gumbel fit parameters")
     response = requests.get(GUMBEL_FIT_DATA, stream=True)
```
```suggestion
    response = requests.get(GUMBEL_FIT_DATA, stream=True)
    if response.status_code != 200:
        raise RuntimeError(
            f"Failed to download Gumbel fit parameters from {GUMBEL_FIT_DATA}. Status code: {response.status_code}"
        )
```

emanuel-schmid commented on climada_petals/hazard/rf_glofas/setup.py:
```diff
@@ -170,37 +97,100 @@ def setup_gumbel_fit(
-    Download dataset of https://doi.org/10.3929/ethz-b-000641667
+    Download dataset of https://doi.org/10.3929/ethz-b-000726304
     """
     LOGGER.debug("Downloading Gumbel fit parameters")
```
If the file is already there, the download could be skipped.
Especially whith the research collection re-organizing this could come in handy.
We'd need a way to check for correctness of the found file.

peanutfun replied:
We usually only check if a file exists, not if it is correct, which is error-prone. We could introduce [pooch](https://www.fatiando.org/pooch/latest/about.html) as a dependency, which automatically verifies the checksum. Would you like to do that?

peanutfun replied:
Clarification: Would **you** like **me** to do that? 😉

peanutfun replied:
I added pooch as dependency for Petals

emanuel-schmid commented on climada_petals/hazard/rf_glofas/transform_ops.py:
```diff
@@ -317,25 +325,32 @@ def download_glofas_discharge(
+        requests=requests,
     )
 
+    if open_mfdataset_kw is False:
```
From the programming perspective this is imo a design flaw.
Any function should always return exactly one return type.
In this case likely files at the cost of having a follow up function that deals with the transformation from files to datasets.

peanutfun replied:
I split the function into one downloading the data and one opening the data.

emanuel-schmid commented on doc/tutorial/climada_hazard_glofas_rf.ipynb:
```diff
@@ -573,149 +555,181 @@
+    "#     preprocess=lambda x: x.sel(time=slice(\"2024-07-10\", \"2024-07-20\")).max(dim=\"time\"),\n",
+    "# )\n",
+    "with xr.open_dataset(\n",
+    "    \"/Users/ldr.riedel/climada/data/cds-download/250627-105456-aa069b92e8690163ba2dc6d2213cb4d9.grib\",\n",
```
```suggestion
    "    f\"{{SYSTEM_DIR}}/cds-download/250627-105456-aa069b92e8690163ba2dc6d2213cb4d9.grib\",\n",
```

peanutfun replied:
Fixed

emanuel-schmid commented on doc/tutorial/climada_hazard_glofas_rf.ipynb:
```diff
@@ -51,34 +41,9 @@
   },
   {
```
`setup_all()` succeeds but only downloads an error page as Gumbel_fits. The request persitently fails with HTTPError 500, "The server is temporarily unable to service your request due to maintenance downtime or capacity problems. Please try again later."
(see suggestion in setup.py)

`setup_all(compute_gumbel_fit=True)` on the other hand fails with HTTPError 404, "Client Error: Not Found for url: https://cds.climate.copernicus.eu/api/retrieve/v1/processes/cems-glofas-historical
process not found \
dataset cems-glofas-historical not found"

peanutfun replied:
The second error is probably because of an outdated config. You will need to update the `.cdsapirc`. The URL has changed. See https://climada-petals--167.org.readthedocs.build/en/167/glofas_rf.html#copernicus-data-store-api-access

peanutfun replied:
The first error is an issue with the research collection. I uploaded the file to Zenodo now.

emanuel-schmid reviewed (changes requested):
I can't get the Tutorial to run. It fails at cell 2.

- if I run it as is, the .grib file is missing
- if I uncomment the lines ahead it fails with 'dataset cems-glofas-historical not found'

i.g. commented lines are fine with me, but I would like to get some infos about why they are there and what I can do with them.

I submit the review for now, because I think it's going to be a lot easier to continue when the tutorial issues are solved or clarified. 
to be continued ...

peanutfun committed:
Apply suggestion from @emanuel-schmid

Co-authored-by: Emanuel Schmid <51439563+emanuel-schmid@users.noreply.github.com>

peanutfun committed:
Update climada_petals/hazard/rf_glofas/setup.py

Co-authored-by: Emanuel Schmid <51439563+emanuel-schmid@users.noreply.github.com>

peanutfun committed:
Introduce 'get_glofas_discharge' and 'open_glofas_discharge'

'get_glofas_discharge' replaces 'download_glofas_discharge', the
latter now only handles downloading itself.

peanutfun committed:
Fix preparation instructions and links

peanutfun committed:
Update .cdsapirc instructions

peanutfun committed:
Use pooch to download GloFAS module static data

peanutfun committed:
Use Zenodo archive for Gumbel fit dataset

peanutfun committed:
Update tutorial

peanutfun committed:
Add pooch to env_docs.yml

peanutfun committed:
Add interpolation fallback option if xesmf is not installed

peanutfun committed:
Update docs with install instructions for xesmf

peanutfun committed:
Remove xesmf from env_climada.yml

peanutfun committed:
Update tests to account for optional xesmf

peanutfun committed:
Update GloFAS river flood tutorial

peanutfun commented:
@elianekobler Please update to the latest version. I uploaded a copy of the fit dataset to Zenodo, and changed the download URL accordingly in bbcc64cf895b5d206b81e2be71bdf74fec4ca3f3

peanutfun commented:
[interpolate_na](https://corteva.github.io/rioxarray/stable/rioxarray.html#rioxarray.raster_array.RasterArray.interpolate_na) only works for 2D and 3D arrays. How to extend to any dim?

peanutfun commented:
Use apply_func like here: https://github.com/MeteoSwiss/meteodata-lab/blob/c9be13a79adbdda8ed7433dbffb40b34156d1465/src/meteodatalab/operators/regrid.py#L308

peanutfun committed:
Add regrid module

peanutfun committed:
ci: Ensure xesmf is installed for not skipping tests

peanutfun committed:
Make spatial interpolation support any number of dimensions
